external/java/src/main/java/ch/epfl/dedis/byzcoin/Proof.java

```diff
@@ -61,38 +75,71 @@ public Instance getInstance() throws CothorityNotFoundException{
     }
 
     /**
-     * accessor for the skipblock assocaited with this proof.
+     * Getter for the skipblock assocaited with this proof.
+     *
      * @return SkipBlock
      */
     public SkipBlock getLatest() {
         return this.latest;
     }
 
     /**
-     * Verifies the proof with regard to the root id. It will follow all
-     * steps and make sure that the hashes work out, starting from the
-     * leaf. At the end it will verify against the root hash to make sure
-     * that the inclusion proof is correct.
+     * Take a skipchain id and verifies that the proof is valid for this
+     * skipchain. It verifies the proof, that the Merkle-root is stored in the
+     * skipblock of the proof and the fact that the skipblock is indeed part of the
+     * skipchain. If all verifications are correct, no exceptions are thrown. It does
+     * not verify whether a certain key/value pair exists in the proof,
+     * use Proof.match or Proof.exists for that.
      *
-     * @param id the skipblock to verify
-     * @return true if all checks verify, false if there is a mismatch in the hashes
-     * @throws CothorityException if something goes wrong
+     * @param scID the skipblock to verify
+     * @throws CothorityCryptoException if something goes wrong
      */
-    public boolean verify(SkipblockId id) throws CothorityException {
-        if (!isByzCoinProof()){
-            return false;
+    public void verify(SkipblockId scID) throws CothorityCryptoException {
+        ByzCoinProto.DataHeader header;
+        try {
+            header = ByzCoinProto.DataHeader.parseFrom(this.latest.getData());
+        } catch (InvalidProtocolBufferException e) {
+            throw new CothorityCryptoException(e.getMessage());
+        }
+        if (!Arrays.equals(this.getRoot(), header.getTrieroot().toByteArray())) {
+            throw new CothorityCryptoException("root of trie is not in skipblock");
+        }
+
+        SkipblockId sbID = null;
+        List<Point> publics = null;
+
+        for (int i = 0; i < this.links.size(); i++) {
+            if (i == 0) {
+                sbID = scID;
+                publics = getPoints(this.links.get(i).getNewRoster().getListList());
+                continue;
+            }
+            ForwardLink l = new ForwardLink(this.links.get(i));
+            if (!l.verify(publics)) {
+                throw new CothorityCryptoException("stored skipblock is not properly evolved from genesis block");
+            }
+            if (!Arrays.equals(l.getFrom().getId(), sbID.getId())) {
+                throw new CothorityCryptoException("stored skipblock is not properly evolved from genesis block");
+            }
+            sbID = l.getTo();
+            try {
+                if (l.getNewRoster() != null) {
+                    publics = getPoints(this.links.get(i).getNewRoster().getListList());
```

Comment:
What on earth is getListList?

Reply:
it's from the generated protobuf code. there's an attribute called `List` and if you want a java `List` of then you gotta call `getListList`